android/ExampleApp/app/src/main/java/com/ingonyama/imp1_aar_example/data/AssetPackManager.kt

```diff
@@ -0,0 +1,164 @@
+import com.google.android.play.core.assetpacks.AssetPackManager
+import com.google.android.play.core.assetpacks.AssetPackManagerFactory
+import com.google.android.play.core.assetpacks.AssetPackState as GoogleAssetPackState
+import com.google.android.play.core.assetpacks.AssetPackStateUpdateListener
+import com.google.android.play.core.assetpacks.model.AssetPackStatus
+import com.ingonyama.imp1_aar_example.Constants
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.map
+
+class AssetPackManager(private val context: Context) {
+    
+    private val assetPackManager: AssetPackManager = AssetPackManagerFactory.getInstance(context)
+    private val assetPackNames: List<String> = com.ingonyama.imp1_aar_example.BuildConfig.ASSET_PACK_NAMES.toList()
+    private val packDownloadProgress = mutableMapOf<String, Int>()
+    private val packDownloadStatus = mutableMapOf<String, Int>()
+    
+    data class AssetPackState(
+        val completedPacks: Int,
+        val totalPacks: Int,
+        val isDownloading: Boolean,
+        val hasFailed: Boolean,
+        val averageProgress: Int,
+        val statusMessage: String,
+        val isReady: Boolean
+    )
+    
+    /**
+     * Checks the status of all zkey asset packs and initiates downloads if needed.
+     */
+    suspend fun checkAssetPacksStatus(): Flow<AssetPackState> = callbackFlow {
+        val packsToDownload = mutableListOf<String>()
+        
+        for (packName in assetPackNames) {
+            val packLocation = assetPackManager.getPackLocation(packName)
+            if (packLocation == null) {
+                // Pack is not available, needs to be downloaded
+                packsToDownload.add(packName)
+                packDownloadStatus[packName] = AssetPackStatus.NOT_INSTALLED
+            } else {
+                // Pack is available
+                packDownloadStatus[packName] = AssetPackStatus.COMPLETED
+                packDownloadProgress[packName] = 100
+            }
+        }
+        
+        // Register the listener to handle updates
+        lateinit var listener: AssetPackStateUpdateListener
+        
+        listener = AssetPackStateUpdateListener { state ->
+            val packName = state.name()
+            if (packName in assetPackNames) {
+                packDownloadStatus[packName] = state.status()
+                updatePackProgress(state)
+                
+                when (state.status()) {
+                    AssetPackStatus.COMPLETED -> {
+                        packDownloadProgress[packName] = 100
+                        trySend(createAssetPackState())
+                        
+                        // Check if all packs are completed
+                        if (packDownloadStatus.values.all { it == AssetPackStatus.COMPLETED }) {
+                            assetPackManager.unregisterListener(listener)
```

Comment:
Manually unregistering in a Flow is problematic and shouldn't be done. It would be better to close (by calling `close()`) the Flow here instead so the `awaitClose` block is called